src/forcing/NetCDFPerFeatureDataProvider.cpp

```diff
@@ -1,7 +1,457 @@
 #ifdef NETCDF_ACTIVE
 #include "NetCDFPerFeatureDataProvider.hpp"
 
+#include <netcdf>
+
+using namespace netCDF;
+using namespace netCDF::exceptions;
+
 std::mutex data_access::NetCDFPerFeatureDataProvider::shared_providers_mutex;
 std::map<std::string, std::shared_ptr<data_access::NetCDFPerFeatureDataProvider>> data_access::NetCDFPerFeatureDataProvider::shared_providers;
 
+namespace data_access {
+
+std::shared_ptr<NetCDFPerFeatureDataProvider> NetCDFPerFeatureDataProvider::get_shared_provider(std::string input_path, time_t sim_start, time_t sim_end, utils::StreamHandler log_s)
+{
+    const std::lock_guard<std::mutex> lock(shared_providers_mutex);
+    std::shared_ptr<NetCDFPerFeatureDataProvider> p;
+    if(shared_providers.count(input_path) > 0){
+        p = shared_providers[input_path];
+    } else {
+        p = std::make_shared<data_access::NetCDFPerFeatureDataProvider>(input_path, sim_start, sim_end, log_s);
+        shared_providers[input_path] = p;
+    }
+    return p;
+}
+
+void NetCDFPerFeatureDataProvider::cleanup_shared_providers()
+{
+    const std::lock_guard<std::mutex> lock(shared_providers_mutex);
+    // First lets try just emptying the map... if all goes well, everything will destruct properly on its own...
+    shared_providers.clear();
+}
+
+NetCDFPerFeatureDataProvider::NetCDFPerFeatureDataProvider(std::string input_path, time_t sim_start, time_t sim_end,  utils::StreamHandler log_s) : log_stream(log_s), value_cache(20),
+    sim_start_date_time_epoch(sim_start),
+    sim_end_date_time_epoch(sim_end)
+{
+    //size_t sizep = 1073741824, nelemsp = 202481;
+    //float preemptionp = 0.75;
+    //nc_set_chunk_cache(sizep, nelemsp, preemptionp);
+
+    //open the file
+    nc_file = std::make_shared<NcFile>(input_path, NcFile::read);
```

Comment:
```suggestion
    nc_file = std::make_shared<netCDF::NcFile>(input_path, netCDF::NcFile::read);
```
This might be the only thing in here requiring `using namespace netCDF;`. I we can get rid of that at the same time, I'd be happy.

Reply:
Hmm... with this change (and removing `using netCDF`) I get:
```
[build] ./ngen-git2/src/forcing/NetCDFPerFeatureDataProvider.cpp:41:76: error: incomplete type ‘netCDF::NcFile’ used in nested name specifier
[build]      nc_file = std::make_shared<netCDF::NcFile>(input_path, netCDF::NcFile::read);
```

Reply:
Did you qualify both? I just tried it, and it works for me